packages/SwingSet/src/kernel/collectionManager.js

```diff
@@ -0,0 +1,891 @@
+    },
+    scalarDurableWeakMapStore: {
+      hasWeakKeys: true,
+      kindID: 0,
+      // eslint-disable-next-line no-use-before-define
+      reanimator: reanimateScalarWeakMapStore,
+      durable: true,
+    },
+    scalarDurableSetStore: {
+      hasWeakKeys: false,
+      kindID: 0,
+      // eslint-disable-next-line no-use-before-define
+      reanimator: reanimateScalarSetStore,
+      durable: true,
+    },
+    scalarDurableWeakSetStore: {
+      hasWeakKeys: true,
+      kindID: 0,
+      // eslint-disable-next-line no-use-before-define
+      reanimator: reanimateScalarWeakSetStore,
+      durable: true,
+    },
+  };
+
+  function prefixc(collectionID, dbEntryKey) {
+    return `vc.${collectionID}.${dbEntryKey}`;
+  }
+
+  function obtainStoreKindID(kindName) {
+    if (storeKindInfoNeedsInitialization) {
+      storeKindInfoNeedsInitialization = false;
+
+      let storeKindIDs = {};
+      const rawTable = syscall.vatstoreGet('storeKindIDTable');
+      if (rawTable) {
+        storeKindIDs = JSON.parse(rawTable);
+      }
+      for (const kind of Object.getOwnPropertyNames(storeKindInfo)) {
+        let kindID = storeKindIDs[kind];
+        if (!kindID) {
+          kindID = allocateExportID();
+          storeKindIDs[kind] = kindID;
+        }
+        storeKindInfo[kind].kindID = kindID;
+        storeKindIDToName.set(`${kindID}`, kind);
+        vrm.registerKind(
+          kindID,
+          storeKindInfo[kind].reanimator,
+          // eslint-disable-next-line no-use-before-define
+          deleteCollection,
+          storeKindInfo[kind].durable,
+        );
+      }
+      syscall.vatstoreSet('storeKindIDTable', JSON.stringify(storeKindIDs));
+    }
+    return storeKindInfo[kindName].kindID;
+  }
+
+  function summonCollectionInternal(
+    initial,
+    label,
+    collectionID,
+    kindName,
+    keySchema = M.any(),
+    valueSchema,
+  ) {
+    const { hasWeakKeys, durable } = storeKindInfo[kindName];
+    const dbKeyPrefix = `vc.${collectionID}.`;
+    let currentGenerationNumber = 0;
+
+    // XXX entryCount needs to be stored persistently, either explicitly or
+    // implicitly, but I'm concerned about the cost.  In the explicit case we
+    // pay for an extra database write each time `init` or `delete` is called,
+    // to increment or decrement a counter stored in the DB.  In the implicit
+    // case we pay a one-time O(n) cost at startup time to count the number of
+    // entries that were there when we last exited (actually, this can be done
+    // lazily instead of at startup, which will save the cost in the likely
+    // common case where nobody ever looks at the entryCount property, but it's
+    // still O(n) when and if it happens).  Neither of these alternatives seems
+    // appetizing, but for now we're going with the lazy, implicit approach.
+    let entryCount;
+
+    if (initial) {
+      entryCount = 0;
+    }
+
+    function ensureEntryCount() {
+      if (entryCount === undefined) {
```

Comment:
Another approach would be to simply not have a `size` method.  We put it in because `Map` and `Set` have a `size` property, but in the case of humongous collections I don't think it actually provides much value -- like, what are you going to do with this number that's useful?

Reply:
yeah that's probably an @erights question, my hunch is that since it has `.keys`, people would expect to also have a `.size`, but it'd be convenient if we could omit it

Reply:
We didn't define a `.size` exactly, but instead

https://github.com/Agoric/agoric-sdk/blob/79c4276da8c856674bd425c54715adec92648c48/packages/store/src/types.js#L194

This is a method rather than an accessor, and so can sensibly invoked with a pipelined eventual method. As @warner observes, since we've got `.keys(keyPattern?, valuePattern?)` and JS Maps and Sets have `.size`, it seemed natural to include this `.getSize(keyPattern?, valuePattern?)` *as a query*.

I agree about the limited utility though. Many uses would me `if (m.getSize() === 0)`, merely because we don't have an `isEmpty(keyPattern?, valuePattern?)` query. This can also be built out of `.keys`, but O(1) rather than O(n).

However, if you search for `getSize` you'll see three uses in agoric-sdk. These seem sensible, so I'm unconvinced yet that we should remove `getSize`.

Reply:
For `getSize` with no pattern arguments, I didn't understand the counterargument to the `entryCount` idea. When is it not (1)? Only when on starting an heir? That doesn't seem too bad. Starting an heir is likely to be very expensive for other reasons anyway.

Reply:
The `entryCount` idea is O(1) in regular operation but O(n) in the number of elements when an upgrade happens.  On the one hand, upgrade is a good time to take the hit if there's going to be an intermittent expensive operation.  On the other hand, if n is a gajillion, that could still suck.

Reply:
On further reflection, it strikes me that it might be reasonable to treat the `entryCount` like a cache.  This PR already puts in place the practice of writing out dirty virtual objects in the virtual object cache when the vat receives a `bringOutYourDead` directive.  This is done in preparation for the upgrade case, so that the state of a vat's data on the disk is complete and consistent at some identifiable moment so that we can shut down the vat for upgrade.  We could use the same occasion to flush the state of `entryCount` of any stores whose entry counts have changed since the last such time.  We have to be a little careful to ensure we don't introduce opportunities for data loss or desynchronization bugs.  This would mean that instead of an additional two database operations for each addition to or deletion from a collection we'd have an additional write once per intentional GC cycle for each changed collection and a single read when starting at upgrade time.  I think this would be quite affordable.

Reply:
Once again closing the loop: @warner and I discussed this and concluded that because the crank buffer effectively acts as a cache, simply storing the entry count as part of the collection's metadata actually would not hurt too badly.  We'd use this to service the dominant case of `getSize` calls with no pattern arguments -- obviously `getSize` _with_ patterns will need to iterate like any other query.   I have implemented this.

We can look at cleverer schemes in the future when and if we find there are performance issues we want to address.